### PR TITLE
Allow using "this" to refer to current token with empty selection

### LINKED ISSUE #1043: Allow using "this" to refer to current token with empty selection
## Use cases
"take this" / "chuck this" / "bring this to bat" with cursor at end of token.  Note that we don't want to auto-expand for implicit destination for a bring command

## Approach
After final modifier stages, but right before running action, run containing token stage if target is empty, weak, and has implicit range

Note that we're rely on the fact that implicit (ie unspoken) target for "bring" automatically has an explicit range to avoid expanding when you say eg "bring air" with an empty cursor

### LINKED ISSUE #1141: Migrate `isImplicit` boolean into its own target type
Instead of
```ts
{
  type: "primitive";
  mark?: Mark;
  modifiers?: Modifier[];
  isImplicit?: boolean;
}
```

Were `isImplicit: true` can't be combined with `mark` or `modifiers` I'd suggest:
```ts
{
  type: "implicit"
}
```

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
- Fixes #1043
- Fixes #1141 
- Depends on https://github.com/cursorless-dev/cursorless/pull/1187

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [x] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)

### CODE REVIEW

File: src/processTargets/processTargets.ts [hunk 1]
```diff
@@ -198,6 +208,10 @@ function processPrimitiveTarget(
     ...context.actionFinalStages,
   ];
 
+  if (!isRangeTarget) {
+    modifierStages.push(new ModifyUnTypedEmptyToTokenStage());
+  }
```
Comment (pokey) [inline src/processTargets/processTargets.ts:213]: How come you skip this on range targets?

File: src/processTargets/modifiers/ModifyIfUntypedStage.ts [hunk 2]
```diff
@@ -1,38 +1,77 @@
...
 import { ProcessedTargetsContext } from "../../typings/Types";
 import getModifierStage from "../getModifierStage";
 import { ModifierStage } from "../PipelineStages.types";
 
-/**
- * Runs {@link ModifyIfUntypedModifier.modifier} if the target has no explicit
- * scope type, ie if {@link Target.hasExplicitScopeType} is `false`.
- */
-export default class ModifyIfUntypedStage implements ModifierStage {
+abstract class ModifyIfBaseStage implements ModifierStage {
```
Comment (pokey) [inline src/processTargets/modifiers/ModifyIfUntypedStage.ts:10]: I'd call this `ConditionalModifierBaseStage`

File: src/processTargets/modifiers/ModifyIfUntypedStage.ts [hunk 3]
```diff
@@ -1,38 +1,77 @@
...
   private get nestedStage() {
     if (this.nestedStage_ == null) {
-      this.nestedStage_ = getModifierStage(this.modifier.modifier);
+      this.nestedStage_ = getModifierStage(this.modifier);
     }
 
     return this.nestedStage_;
   }
+
+  protected abstract modifyIf(target: Target): boolean;
```
Comment (pokey) [inline src/processTargets/modifiers/ModifyIfUntypedStage.ts:42]: I'd call this `shouldModify`

File: src/processTargets/modifiers/ModifyIfUntypedStage.ts [hunk 4]
```diff
@@ -1,38 +1,77 @@
...
-/**
- * Runs {@link ModifyIfUntypedModifier.modifier} if the target has no explicit
- * scope type, ie if {@link Target.hasExplicitScopeType} is `false`.
- */
-export default class ModifyIfUntypedStage implements ModifierStage {
+abstract class ModifyIfBaseStage implements ModifierStage {
   private nestedStage_?: ModifierStage;
 
-  constructor(private modifier: ModifyIfUntypedModifier) {}
+  constructor(private modifier: Modifier, private suppressErrors?: boolean) {}
```
Comment (pokey) [inline src/processTargets/modifiers/ModifyIfUntypedStage.ts:13]: Do you ever use `suppressErrors`?

File: src/test/suite/fixtures/recorded/hatTokenMap/bringHarpAndPointToThisAndStartOfWhaleTakeWhale.yml [hunk 5]
```diff
@@ -35,10 +35,10 @@ initialState:
       start: {line: 0, character: 7}
       end: {line: 0, character: 12}
 finalState:
-  documentContents: hello. hello.world
+  documentContents: hello. .hello
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/hatTokenMap/bringHarpAndPointToThisAndStartOfWhaleTakeWhale.yml:38]: This test needs to be re-recorded to use "just this" instead of "this"

File: src/test/suite/fixtures/recorded/hatTokenMap/bringPointAndHarpToThisAndStartOfWhaleTakeWhale.yml [hunk 6]
```diff
@@ -35,10 +35,10 @@ initialState:
       start: {line: 0, character: 7}
       end: {line: 0, character: 12}
 finalState:
-  documentContents: hello. .helloworld
+  documentContents: hello. hello.
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/hatTokenMap/bringPointAndHarpToThisAndStartOfWhaleTakeWhale.yml:38]: This one too

**Review by pokey — COMMENTED** [review]
Looks good with some minor comments.  Also worth adding tests for:

- [x] "bring this to bat"
- [ ] "bring bat to this"
- [ ] "chuck this"

File: src/processTargets/processTargets.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/processTargets.ts:213]: Because we run this on both the primaries before the range is constructed. I don't think I want `take air past this` or `this slice air` to expand to token?

File: src/processTargets/modifiers/ModifyIfUntypedStage.ts (hunk 4 above)
Comment (AndreasArvidsson) [reply src/processTargets/modifiers/ModifyIfUntypedStage.ts:13]: Yes the new modifier passes it in the constructor

Comment (pokey) [reply src/processTargets/modifiers/ModifyIfUntypedStage.ts:13]: Ah right I see now; that line is a bit hard to read with the inline obj

Comment (pokey) [inline src/processTargets/modifiers/ModifyIfUntypedStage.ts:13]: Also, I think I'd call the nested modifier `nestedModifier`, because usually `modifier` refers to the modifier for which we are constructing a stage

File: src/processTargets/processTargets.ts (hunk 1 above)
Comment (pokey) [reply src/processTargets/processTargets.ts:213]: Huh.  This one feels a bit inconsistent, but I guess I agree.  Are there tests for this case?

File: src/processTargets/modifiers/ModifyIfUntypedStage.ts [hunk 7]
```diff
@@ -1,38 +1,77 @@
...
+  }
+}
+
+/**
+ * Tries to convert target into token if:
+ * {@link Target.hasExplicitScopeType} is `false` and
+ * {@link Target.hasExplicitRange} is `false` and
+ * {@link Target.contentRange.isEmpty} is `true`.
+ */
+export class ModifyUnTypedEmptyToTokenStage extends ModifyIfBaseStage {
```
Comment (pokey) [inline src/processTargets/modifiers/ModifyIfUntypedStage.ts:65]: Capitalisation here is inconsistent with `modifyIfUntyped`: `UnTyped` should be `Untyped`.  Actually this name is a bit wonky overall.  Maybe `ContainingTokenIfUntypedEmptyStage`?

File: src/processTargets/processTargets.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/processTargets.ts:213]: There is now :D

**Comment by AndreasArvidsson** [comment]
@pokey Ready!

**Comment by pokey** [comment]
Thinking about it more, I don't think we should have special cases for ranges.  It is confusing if "bring this to bat" behaves differently from "take this past bat".  Happy to discuss at meet-up if you prefer the special case for ranges

File: src/processTargets/modifiers/ConditionalModifierStages.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  Modifier,
+  ModifyIfUntypedModifier,
+} from "../../typings/targetDescriptor.types";
+import { ProcessedTargetsContext } from "../../typings/Types";
+import getModifierStage from "../getModifierStage";
+import { ModifierStage } from "../PipelineStages.types";
+
+abstract class ConditionalModifierBaseStage implements ModifierStage {
+  private nestedStage_?: ModifierStage;
+  protected suppressErrors = false;
```
Comment (pokey) [inline src/processTargets/modifiers/ConditionalModifierStages.ts:12]: This feels orthogonal to conditional modifier.  Should really just use #1015 once we have that.  I'm ok with leaving this as is for now, as that machinery isn't there yet, but prob worth adding a checkbox to #1015 to use that machinery here

**Comment by AndreasArvidsson** [comment]
> Thinking about it more, I don't think we should have special cases for ranges. It is confusing if "bring this to bat" behaves differently from "take this past bat". Happy to discuss at meet-up if you prefer the special case for ranges

I'm not sure either. Lets every discussion.

**Comment by pokey** [comment]
Update based on discussion:

- [ ] No special cases for range targets
- [ ] Make sure that "take past air" uses implicit target for anchor
- [ ] Update tests, potentially splitting some of them into up to three tests:
   - [ ] using `"just"` to get old behaviour
   - [ ] using implicit target to get old behaviour
   - [ ] new behaviour

**Comment by AndreasArvidsson** [comment]
@pokey ready!

File: src/processTargets/processTargets.ts [hunk 9]
```diff
@@ -229,7 +231,13 @@ export function processModifierStages(
...
 
 function calcIsReversed(anchor: Target, active: Target) {
-  return anchor.contentRange.start.isAfter(active.contentRange.start);
+  if (anchor.contentRange.start.isAfter(active.contentRange.start)) {
+    return true;
+  }
+  if (anchor.contentRange.start.isBefore(active.contentRange.start)) {
+    return false;
+  }
+  return anchor.contentRange.end.isAfter(active.contentRange.end);
```
Comment (pokey) [inline src/processTargets/processTargets.ts:243]: how come this had to change?

File: src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastEndOfToken.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  marks: {}
+finalState:
+  documentContents: hello th
+  selections:
+    - anchor: {line: 0, character: 8}
+      active: {line: 0, character: 8}
+fullTargets: [{type: range, excludeAnchor: false, excludeActive: false, rangeType: continuous, anchor: {type: primitive, mark: {type: cursor}, modifiers: [{type: toRawSelection}]}, active: {type: primitive, mark: {type: cursor}, modifiers: [{type: containingScope, scopeType: {type: token}}], positionModifier: {type: position, position: end}}}]
+ide:
+  messages:
+    - {type: warning, id: deprecatedPositionInference}
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastEndOfToken.yml:36]: how come we're using this deprecated command form?  presumably we should switch to "past end of its token"?

File: src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastStartOfToken.yml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  marks: {}
+finalState:
+  documentContents: hello ere
+  selections:
+    - anchor: {line: 0, character: 6}
+      active: {line: 0, character: 6}
+fullTargets: [{type: range, excludeAnchor: false, excludeActive: false, rangeType: continuous, anchor: {type: primitive, mark: {type: cursor}, modifiers: [{type: toRawSelection}]}, active: {type: primitive, mark: {type: cursor}, modifiers: [{type: containingScope, scopeType: {type: token}}], positionModifier: {type: position, position: start}}}]
+ide:
+  messages:
+    - {type: warning, id: deprecatedPositionInference}
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastStartOfToken.yml:36]: as above

File: src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
 command:
   version: 1
-  spokenForm: take past start of token
-  action: setSelection
+  spokenForm: clear past start of token
+  action: clearAndSetSelection
   targets:
     - type: range
-      start: {type: primitive}
+      start: {type: primitive, isImplicit: true}
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml:8]: do we need to do a migration?  what happens when they issue the command with new extension / old Talon?  Does this no longer work as it did before?

File: src/test/suite/fixtures/recorded/compoundTargets/clearThisSliceBatt.yml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    - anchor: {line: 0, character: 0}
+      active: {line: 0, character: 1}
+  marks:
+    default.b:
+      start: {line: 1, character: 0}
+      end: {line: 1, character: 2}
+finalState:
+  documentContents: |-
+    a
+    b
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/compoundTargets/clearThisSliceBatt.yml:32]: huh.  is this what we want?  seems inconsistent

**Review by pokey — COMMENTED** [review]
Looks pretty good.  I fixed merge conflicts and tweaked comments, but see questions below

File: src/processTargets/processTargets.ts (hunk 9 above)
Comment (AndreasArvidsson) [reply src/processTargets/processTargets.ts:243]: This was incorrectly implemented, but only for very niche cases. If I remember correctly it had to do with reverse selection `this past start of token` was inconsistent with `this past end of token`

File: src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastEndOfToken.yml (hunk 10 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/compoundTargets/clearJustThisPastEndOfToken.yml:36]: Doesn't matter for this test.

File: src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml (hunk 12 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml:8]: Should work as before

File: src/test/suite/fixtures/recorded/compoundTargets/clearThisSliceBatt.yml (hunk 13 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/compoundTargets/clearThisSliceBatt.yml:32]: The selection is non empty and therefore does not expand

File: src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml (hunk 12 above)
Comment (pokey) [reply src/test/suite/fixtures/recorded/compoundTargets/clearPastStartOfToken.yml:8]: Needs a migration 😕.  I can knock it out

**Comment by pokey** [comment]
Unfortunately, [`"square switch past whale"`](https://github.com/cursorless-dev/cursorless/blob/main/src/test/suite/fixtures/recorded/implicitExpansion/squareSwitchPastWhale.yml) breaks, because this PR makes that command equivalent to `"square switch just this past whale"`, which means that the `square switch` action won't do implicit expansion on `this` because of the `just` 😕.

Not immediately obvious how to fix this one cc/ @AndreasArvidsson

**Comment by AndreasArvidsson** [comment]
Isn't that the desired behavior? If you say this it will expand but implicit targets do not?

**Comment by pokey** [comment]
I guess maybe it's ok?  It's causing 4 CI failures...

**Review by pokey — APPROVED** [review]
Ok I think this one is good to go; I'll give you a chance to have a quick look if you get a minute @AndreasArvidsson

**Comment by pokey** [comment]
Update from meet-up:
- Let's guard this one with a setting: `cursorless.experimental.featureFlags.emptyCursorTokenExpansion`
- Let's add tests that exercise this setting

**Comment by AndreasArvidsson** [comment]
@pokey Looks good

File: src/core/commandVersionUpgrades/upgradeV3ToV4/upgradeV3ToV4.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      };
+    }
+    case "list": {
+      const { elements, ...rest } = target;
+
+      return {
+        ...rest,
+        elements: elements.map(upgradeTarget) as [
+          PartialPrimitiveTargetDescriptor | PartialRangeTargetDescriptor,
+        ],
```
Comment (auscompgeek) [inline src/core/commandVersionUpgrades/upgradeV3ToV4/upgradeV3ToV4.ts:42]: This should be an array rather than a 1-tuple, right?

**Comment by pokey** [comment]
I found myself wanting this one again today; I say let's ship it without feature flag.  If we decide it's a bad idea down the road, we can always:

1. move it behind a feature flag, then
2. have default behaviour (no feature flag) be to show a warning, pointing them to doc for enabling feature flag

File: src/core/commandVersionUpgrades/upgradeV3ToV4/upgradeV3ToV4.ts (hunk 14 above)
Comment (pokey) [reply src/core/commandVersionUpgrades/upgradeV3ToV4/upgradeV3ToV4.ts]: Nice catch 😊

**Review by pokey — APPROVED** [review]
